pkg/controller/direct/monitoring/maputils.go

```diff
@@ -0,0 +1,179 @@
+package monitoring
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/durationpb"
+)
+
+type MapContext struct {
+	errs []error
+}
+
+func (c *MapContext) Errorf(msg string, args ...interface{}) {
+	c.errs = append(c.errs, fmt.Errorf(msg, args...))
+}
+
+func (c *MapContext) Err() error {
+	return errors.Join(c.errs...)
+}
+
+func Slice_ToProto[T, U any](ctx *MapContext, in []T, mapper func(ctx *MapContext, in *T) *U) []*U {
+	if in == nil {
+		return nil
+	}
+
+	outSlice := make([]*U, 0, len(in))
+	for _, inItem := range in {
+		outItem := mapper(ctx, &inItem)
+		outSlice = append(outSlice, outItem)
+	}
+	return outSlice
+}
+
+func Slice_FromProto[T, U any](ctx *MapContext, in []*T, mapper func(ctx *MapContext, in *T) *U) []U {
+	if in == nil {
+		return nil
+	}
+
+	outSlice := make([]U, 0, len(in))
+	for _, inItem := range in {
+		outItem := mapper(ctx, inItem)
+		outSlice = append(outSlice, *outItem)
+	}
+	return outSlice
+}
+
+type ProtoEnum interface {
+	~int32
+	Descriptor() protoreflect.EnumDescriptor
+}
+
+func Enum_ToProto[U ProtoEnum](ctx *MapContext, in *string) U {
```

Comment:
big nit: Going to leave it up to you but could we rename `ctx` to `mapCtx` or something that doesn't make me think of the `context.Context` one 😛

Reply:
A much better name, thank you!